lib/Core/BuildEngine.cpp

```diff
@@ -959,12 +1035,22 @@ class BuildEngineImpl : public BuildDBDelegate {
         // Report the status change.
         ruleInfo->rule->updateStatus(buildEngine, Rule::StatusKind::IsComplete);
 
+        // Add all dependencies with barriers between the batches
+        for (auto &dependencies : taskInfo->sequencedDependencyBatches) {
+          ruleInfo->result.dependencies.addBarrier();
+          ruleInfo->result.dependencies.append(dependencies.second);
+        }
+        taskInfo->sequencedDependencyBatches.clear();
+
         // Add all of the task's discovered dependencies.
         //
         // FIXME: We could audit these dependencies at this point to verify that
         // they are not keys for rules which have not been run, which would
         // indicate an underspecified build (e.g., a generated header).
-        ruleInfo->result.dependencies.append(taskInfo->discoveredDependencies);
+        if (!taskInfo->discoveredDependencies.empty()) {
+          ruleInfo->result.dependencies.addBarrier();
+          ruleInfo->result.dependencies.append(taskInfo->discoveredDependencies);
+        }
 
```

Comment:
@erikolofsson: The builds are not triggering correctly because this piece of code doesn't seem to be executed in case of build failures. In the subsequent builds, the `result.dependencies` is empty, resulting in `ruleDoesNotNeedToRun`. A not-so-nice solution would be to re-add the following near L904
```
request.taskInfo->forRuleInfo->result.dependencies.push_back(
            request.inputRuleInfo->keyID, request.orderOnly, request.singleUse);
```

Ideally, we should move the `sequencedDependencyBatches` inside `Result` but I find it risky since `Result` is persisted inside the Sqlite DB.

Reply:
Thanks. I refactored so `sequencedDependencyBatches` isn't needed and it adds to the result directly (See [branch](https://github.com/erikolofsson/swift-llbuild/tree/incremental-parallel-20248283-2)). I will be doing some longer term testing and if it holds up I will push the fix here.

Reply:
Cool. We'll test this out here at Uber as well 😃

Reply:
I noticed that the incremental build wasn't parallell when it should have been so I fixed this by recording the batch sequence at the time the input was requested instead of when it was popped off the queue and pushed the fix to the [branch](https://github.com/erikolofsson/swift-llbuild/tree/incremental-parallel-20248283-2).